問題：撰寫一Python程式，為使用者輸入年份（可能是西元或民國）找尋其相對應的十二生肖。十二生肖是以十二年為一週期，每一年分別以一種動物來表示猴（monkey）、雞（rooster）、狗(dog)、豬(pig)、鼠(rat)、牛(ox)、虎(tiger)、兔(rabbit)、龍(dragon)、蛇（snake）、馬(horse)，以及羊(sheep) 。

請依序回答下面問題：

1. 0.1的二進位表示法為何? 使用markdown語法，列出計算過程和理由，並描述看到的現象。

2. 如果是無限多位，Python如何解決？提示：IEEE 754、四捨六入五成雙

3. Python如何知道浮點數的精度？請問執行以下程式碼:
```python
import sys
print(sys.float_info)
```
並說明輸出結果的意思。

４. 在數學上0.1+0.2=0.3，可是下面的Python程式的答案好像不是0.3。怎麼解決？
```python
print(0.1+0.2)
```

5. 你的方法，對於一般兩個浮點數加法，在科學或金融上是好方法嗎？請思考。



In [4]:
import sys
print(sys.float_info)

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)


In [5]:
print(0.1 + 0.2)

0.30000000000000004


In [6]:
round(0.1 + 0.2, 1)

0.3

1.將十進位小數 0.1 轉換成二進位：
因此可得到：
0.1₁₀ = 0.0001100110011001100110011...₂
其中：
0011
會不斷重複出現，因此是無限循環二進位小數。

2.Python 的 float 採用 IEEE 754 Double Precision（雙精度浮點數）。
結構:
1 bit 符號位
11 bits 指數位
52 bits 尾數位

3.輸出結果會顯示 Python 浮點數（float）的相關資訊，例如最大值、最小值、精度及誤差範圍。
 max:float 可表示的最大數值。
 min:float 可表示的最小正規化正數。
 dig:約能正確表示 15 位十進位有效數字。
 mant_dig = 53：尾數精度為 53 位元。
 epsilon ≈ 2.22 × 10⁻¹⁶：表示浮點數能分辨的最小差距。

4.原因：
0.1 與 0.2 都是近似值。
相加後誤差被保留下來。

5.科學計算
math.isclose()
浮點誤差無法完全避免。
科學計算通常允許極小誤差。
判斷是否「足夠接近」比直接比較更合理。

..金融計算
不建議使用float
會產生累積誤差。
金額計算需要精確結果。
建議使用 Decimal

In [3]:
zodiac = [
    "monkey",
    "rooster",
    "dog",
    "pig",
    "rat",
    "ox",
    "tiger",
    "rabbit",
    "dragon",
    "snake",
    "horse",
    "sheep"
]

year = int(input("請輸入年份："))

# 若輸入為民國年，轉換成西元年
if year < 1912:
    year = year + 1911

print("生肖為：", zodiac[year % 12])

生肖為： dog
